ibis/mapd/compiler.py

```diff
@@ -0,0 +1,220 @@
+    """
+
+    """
+    _join_names = {
+        ops.InnerJoin: 'JOIN',
+        ops.LeftJoin: 'LEFT JOIN'
+    }
+
+    def get_result(self):
+        # Got to unravel the join stack; the nesting order could be
+        # arbitrary, so we do a depth first search and push the join tokens
+        # and predicates onto a flat list, then format them
+        op = self.expr.op()
+
+        if isinstance(op, ops.Join):
+            self._walk_join_tree(op)
+        else:
+            self.join_tables.append(self._format_table(self.expr))
+
+        buf = StringIO()
+        buf.write(self.join_tables[0])
+        for jtype, table, preds in zip(self.join_types, self.join_tables[1:],
+                                       self.join_predicates):
+            buf.write('\n')
+            buf.write(util.indent('{0} {1}'.format(jtype, table), self.indent))
+
+            if len(preds):
+                buf.write('\n')
+                fmt_preds = map(self._format_predicate, preds)
+                fmt_preds = util.indent('USING ' + ', '.join(fmt_preds),
+                                        self.indent * 2)
+                buf.write(fmt_preds)
+
+        return buf.getvalue()
+
+    def _validate_join_predicates(self, predicates):
+        for pred in predicates:
+            op = pred.op()
+            if not isinstance(op, ops.Equals):
+                raise com.TranslationError(
+                    'Non-equality join predicates are not supported'
+                )
+
+            left_on, right_on = op.args
+            if left_on.get_name() != right_on.get_name():
+                raise com.TranslationError(
+                    'Joining on different column names is not supported'
+                )
+
+    def _format_predicate(self, predicate):
+        column = predicate.op().args[0]
+        return column.get_name()
+
+    def _quote_identifier(self, name):
+        return name
+
+
+class MapDExprTranslator(compiles.ExprTranslator):
+    """
+
+    """
+    _registry = mapd_ops._operation_registry
+    context_class = MapDQueryContext
+
+    def name(self, translated, name, force=True):
+        return mapd_ops._name_expr(translated, name)
+
+
+class MapDDialect(compiles.Dialect):
+    """
+
+    """
+    translator = MapDExprTranslator
+
+
+dialect = MapDDialect
+compiles = MapDExprTranslator.compiles
+rewrites = MapDExprTranslator.rewrites
+
+compiles(ops.Distance, mapd_ops.distance)
+
+mapd_reg = mapd_ops._operation_registry
+
+_add_methods(
+    ir.NumericValue, dict(
+        conv_4326_900913_x=_unary_op(
+            'conv_4326_900913_x', mapd_ops.Conv_4326_900913_X
+        ),
+        conv_4326_900913_y=_unary_op(
+            'conv_4326_900913_y', mapd_ops.Conv_4326_900913_Y
+        ),
+        truncate=_binop_expr(
+            'truncate', mapd_ops.NumericTruncate
+        )
+    )
+)
+
+_add_methods(
+    ir.StringValue, dict(
+        byte_length=_unary_op('length', mapd_ops.ByteLength)
```

Comment:
I think this should be moved to `ibis/expr/api.py`

Reply:
Indeed, having a different API for bytes vs strings isn't great, but one may also want to compute the number of bytes in a string value as well as the semantic length of the string in human-readable characters